glom/grouping.py

```diff
@@ -0,0 +1,292 @@
+    raises exception on non-numeric value
+
+    >>> glom([1, 2, 3], Group(Avg()))
+    2.0
+    """
+    __slots__ = ()
+
+    def agg(self, target, tree):
+        try:
+            avg_acc = tree[self]
+        except KeyError:
+            # format is [sum, count]
+            avg_acc = tree[self] = [0.0, 0]
+        avg_acc[0] += target
+        avg_acc[1] += 1
+        return avg_acc[0] / avg_acc[1]
+
+
+class Sum(object):
+    """
+    takes the sum of all values;
+    raises exception on values incompatible with addition operator
+
+    >>> glom([1, 2, 3], Group(Sum()))
+    6
+    """
+    __slots__ = ()
+
+    def agg(self, target, tree):
+        if self not in tree:
+            tree[self] = 0
+        tree[self] += target
+        return tree[self]
+
+
+class Max(object):
+    """
+    takes the maximum of all values;
+    raises exception on values that are not comparable
+
+    >>> glom([1, 2, 3], Group(Max()))
+    3
+    """
+    __slots__ = ()
+
+    def agg(self, target, tree):
+        if self not in tree or target > tree[self]:
+            tree[self] = target
+        return tree[self]
+
+
+class Min(object):
+    """
+    takes the minimum of all values;
+    raises exception on values that are not comparable
+
+    >>> glom([1, 2, 3], Group(Min()))
+    1
+    """
+    __slots__ = ()
+
+    def agg(self, target, tree):
+        if self not in tree or target < tree[self]:
+            tree[self] = target
+        return tree[self]
+
+
+class Count(object):
+    """
+    takes a count of how many values occurred
+
+    >>> glom([1, 2, 3], Group(Count()))
+    3
+    """
+    __slots__ = ()
+
+    def agg(self, target, tree):
+        if self not in tree:
+            tree[self] = 0
+        tree[self] += 1
+        return tree[self]
+
+
+# TODO: is this a good idea or better to use random.sample()?
+class Sample(object):
+    """
+    takes a random sample of the values
+
+    >>> glom([1, 2, 3], Group(Sample(2)))  # doctest: +SKIP
+    [1, 3]
+    >>> glom([1, 2, 3], Group(Sample(2)))  # doctest: +SKIP
+    [3, 2]
+    """
+    __slots__ = ('size',)
+
+    def __init__(self, size):
+        self.size = size
+
+    def agg(self, target, tree):
+        # simple reservoir sampling scheme
+        # https://en.wikipedia.org/wiki/Reservoir_sampling#Simple_algorithm
+        if self not in tree:
+            tree[self] = [0, []]
+        num_seen, sample = tree[self]
+        if len(sample) < self.size:
+            sample.append(target)
+        else:
+            pos = random.randint(0, num_seen)
+            if pos < self.size:
+                sample[pos] = target
+        tree[self][0] += 1
+        return sample
+
+
+class Limit(object):
+    """
+    limits the number of values passed to sub-accumulator
+
+    >>> glom([1, 2, 3], Group(T))
+    3
+    >>> glom([1, 2, 3], Group(Limit(1)))
+    1
+    """
+    __slots__ = ('n', 'subspec')
+
+    def __init__(self, n, subspec=T):
```

Comment:
T default here I feel is a bad idea -- for 90% of specs T is a safe/sane default

however, for Limit, T basically means "Last" -- which is randomly specific in the same way that having "Last" or "Avg" be the default would be

Reply:
Hmm, maybe I'm misunderstanding, isn't this a transform applied to each element that's actually selected? Like, `Limit(1)` resulted in `1` being selected, not `3`, so how does it result in Last? Are you referring to `T` being the default in `Group()` bc that's something else.

Which reminds me, what does the dict in this do:
```python
if self not in tree:
   tree[self] = [0, {}]
```

Reply:
Limit only works inside group; yes it is applied to each element, but within a group only the last element result is kept

Reply:
```python
glom([1, 2, 3, 4, 5], Group(Limit(3)))
# 3
```

vs
```python
glom([1, 2, 3, 4, 5], Limit(3))
# [1, 2, 3]
```

Reply:
(the second one doesn't work)